services/filemanager/s3manager.go

```diff
@@ -190,27 +193,27 @@ func (manager *S3Manager) getSession(ctx context.Context) (*session.Session, err
 	} else {
 		region = *manager.Config.Region
 	}
+
 	var sess *session.Session
-	if manager.Config.AccessKeyID == "" || manager.Config.AccessKey == "" {
-		pkgLogger.Debug("Credentials not found in the destination's config. Using the host credentials instead")
-		sess = session.Must(session.NewSession(&aws.Config{
-			Region:                        aws.String(region),
-			CredentialsChainVerboseErrors: aws.Bool(true),
-			Endpoint:                      manager.Config.Endpoint,
-			S3ForcePathStyle:              manager.Config.S3ForcePathStyle,
-			DisableSSL:                    manager.Config.DisableSSL,
-		}))
-	} else {
-		pkgLogger.Debug("Credentials found in the destination's config.")
-		sess = session.Must(session.NewSession(&aws.Config{
-			Region:                        aws.String(region),
-			Credentials:                   credentials.NewStaticCredentials(manager.Config.AccessKeyID, manager.Config.AccessKey, ""),
-			CredentialsChainVerboseErrors: aws.Bool(true),
-			Endpoint:                      manager.Config.Endpoint,
-			S3ForcePathStyle:              manager.Config.S3ForcePathStyle,
-			DisableSSL:                    manager.Config.DisableSSL,
-		}))
+	awsConfig := aws.Config{
+		Region:                        aws.String(region),
+		CredentialsChainVerboseErrors: aws.Bool(true),
+		Endpoint:                      manager.Config.Endpoint,
+		S3ForcePathStyle:              manager.Config.S3ForcePathStyle,
+		DisableSSL:                    manager.Config.DisableSSL, // TODO: get rid of this options as its risky
+	}
+	if manager.Config.AccessKey != "" && manager.Config.AccessKeyID != "" {
+		awsConfig.Credentials = credentials.NewStaticCredentials(manager.Config.AccessKeyID, manager.Config.AccessKey, "")
+	} else if manager.Config.IAMRoleARN != "" {
+		stsSession := session.Must(session.NewSession(&awsConfig))
+		awsConfig.Credentials = stscreds.NewCredentials(stsSession, manager.Config.IAMRoleARN, func(p *stscreds.AssumeRoleProvider) {
+			p.ExternalID = aws.String(manager.Config.ExternalID)
+			p.RoleSessionName = crossAccountSTSRoleSessionName
+		})
 	}
+
+	sess = session.Must(session.NewSession(&awsConfig))
+
 	return sess, nil
```

Comment:
Given that this method can return errors, wouldn't it make more sense to avoid panicking with must and just propagate the error up?

Reply:
https://github.com/rudderlabs/rudder-server/pull/2251